promql/engine.go

```diff
@@ -3871,3 +4010,77 @@ func (ev *evaluator) gatherVector(ts int64, input Matrix, output Vector, bufHelp
 
 	return output, bufHelpers
 }
+
+// scalarValueOf calculates the scalar value of an expression in float64.
+// Use case for this is to be able to calculate offset durations,
+// timestamp, range duration, etc. from scalar expressions.
+func (ev *evaluator) scalarValueOf(e parser.Expr) (float64, error) {
+	if e == nil {
+		return 0, nil
+	}
+
+	if ev.querier != nil {
+		panic("evaulator does not allow querier for time expressions")
+	}
+
+	// Shortcut for number literals.
+	if nl, ok := e.(*parser.NumberLiteral); ok {
+		return nl.Val, nil
+	}
+
+	// We can only use scalars as timestamp or duration.
+	if e.Type() != parser.ValueTypeScalar {
+		panic("parser should have discarded non scalar time expression")
+	}
+
+	// TODO(krajorama): would we ever get annotations here?
```

Comment:
Currently this could happen in an extremely contrived case like `(scalar(limit_ratio(-2, vector(0))))`